ambry-tools/src/main/java/com.github.ambry.tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,946 @@
+    // Creating PutThread
+    Thread putThread = new Thread(
+        new PutThread("Producer", putGetHelper, sharedQueue, maxPutCount, currentPutCount, maxParallelPutCount,
+            sleepTimeBetweenBatchPutsInMs, enableVerboseLogging, threadLocalRandom));
+
+    logger.trace("Initiating Consumer thread ");
+    // Creating Get thread
+    Thread getThread = new Thread(
+        new GetThread("Consumer", putGetHelper, sharedQueue, maxGetCountPerBlob, maxPutCount, getCompletedCount,
+            burstCountGet, parallelGetCount, sleepTimeBetweenBatchGetsInMs, deleteOnExit, enableVerboseLogging,
+            threadLocalRandom, deletedBlobIds));
+
+    Thread deleteThread = null;
+    if (deleteOnExit) {
+      logger.trace("Initiating Delete thread ");
+      DeleteThread deleteBlobThread =
+          new DeleteThread("DeleteThread", putGetHelper, maxPutCount, deletedBlobIds, enableVerboseLogging);
+      deleteThread = new Thread(deleteBlobThread);
+      deleteThread.start();
+    }
+
+    logger.trace("Starting Producer thread ");
+    putThread.start();
+    logger.trace("Starting Consumer thread ");
+    getThread.start();
+
+    putThread.join();
+    getThread.join();
+
+    if (deleteThread != null) {
+      logger.trace("Starting Delete thread ");
+      deleteThread.join();
+    }
+
+    logger.trace("Invoking shutdown after completing the test suite ");
+    // shutting down the PutGetHelper class to release any resources held up
+    putGetHelper.shutdown();
+  }
+
+  /**
+   * Put threads to assist in producing blobs or uploading blobs to ambry
+   */
+  class PutThread implements Runnable {
+    private final String threadName;
+    private final PutGetHelper producerJob;
+    private final ArrayList<BlobAccessInfo> blobInfos;
+    private final int maxPutCount;
+    private final AtomicInteger currentPutCount;
+    private final int maxParallelRequest;
+    private final long sleepTimeBetweenBatchPutsInMs;
+    private final boolean enableVerboseLogging;
+    private final ThreadLocalRandom threadLocalRandom;
+
+    public PutThread(String threadName, PutGetHelper producerJob, ArrayList<BlobAccessInfo> blobInfos,
+        final int maxPutCount, AtomicInteger currentPutCount, int maxParallelRequest,
+        long sleepTimeBetweenBatchPutsInMs, boolean enableVerboseLogging, ThreadLocalRandom threadLocalRandom) {
+      this.threadName = threadName;
+      this.producerJob = producerJob;
+      this.blobInfos = blobInfos;
+      this.maxPutCount = maxPutCount;
+      this.currentPutCount = currentPutCount;
+      this.maxParallelRequest = maxParallelRequest;
+      this.sleepTimeBetweenBatchPutsInMs = sleepTimeBetweenBatchPutsInMs;
+      this.enableVerboseLogging = enableVerboseLogging;
+      this.threadLocalRandom = threadLocalRandom;
+    }
+
+    @Override
+    public void run() {
+      while (currentPutCount.get() < maxPutCount) {
+        try {
+          int burstCount = threadLocalRandom.nextInt(maxParallelRequest) + 1;
```

Comment:
Puts happen in multiple cycles. On every cycle, we make N no of burst puts. N is a random value upto a max of "maxParallelRequest". So, such batch puts will continue until we have made "totalPutCount" of PUTs.